pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,360 @@
+				log.Debugf("HTTPMatchRequests conflict: root %v, delegate %v", rootMatch, subMatch)
+				continue
+			}
+			// merge HTTPMatchRequest
+			out = append(out, mergeHTTPMatchRequest(rootMatch, subMatch))
+			foundMatch = true
+		}
+		if !foundMatch {
+			return nil, true
+		}
+	}
+	if len(out) == 0 {
+		conflict = true
+	}
+	return
+}
+
+func mergeHTTPMatchRequest(root, delegate *networking.HTTPMatchRequest) *networking.HTTPMatchRequest {
+	out := *delegate
+	if out.Name == "" {
+		out.Name = root.Name
+	}
+	if out.Uri == nil {
+		out.Uri = root.Uri
+	}
+	if out.Scheme == nil {
+		out.Scheme = root.Scheme
+	}
+	if out.Method == nil {
+		out.Method = root.Method
+	}
+	if out.Authority == nil {
+		out.Authority = root.Authority
+	}
+	// headers
+	if len(root.Headers) > 0 || len(delegate.Headers) > 0 {
+		out.Headers = make(map[string]*networking.StringMatch)
+	}
+	for k, v := range root.Headers {
+		out.Headers[k] = v
+	}
+	for k, v := range delegate.Headers {
+		out.Headers[k] = v
+	}
+	// withoutheaders
+	if len(root.WithoutHeaders) > 0 || len(delegate.WithoutHeaders) > 0 {
+		out.WithoutHeaders = make(map[string]*networking.StringMatch)
+	}
+	for k, v := range root.WithoutHeaders {
+		out.WithoutHeaders[k] = v
+	}
+	for k, v := range delegate.WithoutHeaders {
+		out.WithoutHeaders[k] = v
+	}
+	// queryparams
+	if len(root.QueryParams) > 0 || len(delegate.QueryParams) > 0 {
+		out.QueryParams = make(map[string]*networking.StringMatch)
+	}
+	for k, v := range root.QueryParams {
+		out.QueryParams[k] = v
+	}
+	for k, v := range delegate.QueryParams {
+		out.QueryParams[k] = v
+	}
+	// sourcelabels
+	if len(root.SourceLabels) > 0 || len(delegate.SourceLabels) > 0 {
+		out.SourceLabels = make(map[string]string)
+	}
+	for k, v := range root.SourceLabels {
+		out.SourceLabels[k] = v
+	}
+	for k, v := range delegate.SourceLabels {
+		out.SourceLabels[k] = v
+	}
+	if len(out.Gateways) == 0 {
+		out.Gateways = root.Gateways
+	}
+	if out.SourceNamespace == "" {
+		out.SourceNamespace = root.SourceNamespace
+	}
+	if out.Port == 0 {
+		out.Port = root.Port
+	}
+	return &out
+}
+
+func hasConflict(root, leaf *networking.HTTPMatchRequest) bool {
+	roots := []*networking.StringMatch{root.Uri, root.Scheme, root.Method, root.Authority}
+	leaves := []*networking.StringMatch{leaf.Uri, leaf.Scheme, leaf.Method, leaf.Authority}
+	for i := range roots {
+		if stringMatchConflict(roots[i], leaves[i]) {
+			return true
+		}
+	}
+	// header conflicts
+	for key, leafHeader := range leaf.Headers {
+		if stringMatchConflict(root.Headers[key], leafHeader) {
+			return true
+		}
+	}
+
+	// without headers
+	for key, leafValue := range leaf.WithoutHeaders {
+		if stringMatchConflict(root.Headers[key], leafValue) {
+			return true
+		}
+	}
+
+	// query params conflict
+	for key, value := range leaf.QueryParams {
+		if stringMatchConflict(root.QueryParams[key], value) {
+			return true
+		}
+	}
+
+	// source labels
+	for key, leafValue := range leaf.SourceLabels {
```

Comment:
we can reject in validation. Source labels are applicable only for sidecars. Not for gateways where this delegation is done.

Reply:
ic